wnw.py

```diff
@@ -0,0 +1,171 @@
+"""Utilities for the word-nonword study"""
+
+import os.path as op
+
+
+WNW_DEFAULT_ROOT = "/data/NIMH_SFIM/handwerkerd/DenoisingPilot"
+WNW_SUBJECTS = [1, 3, 4]
+WNW_MODALITIES = ["func", "anat", "T1w"]
+WNW_RUNS = [1, 2]
+WNW_ACQUISITIONS = ["a", "b", "c"]
+
+WNW_ECHOES = {
+    "a": [1, 2, 3],
+    "b": [1, 2, 3, 4, 5],
+    "c": [1, 2, 3],
+}
+
+
+class WnwProject:
+    """An object to represent the underlying word-nonword data structure
+
+    Attributes
+    ----------
+    _root: str
+        The root directory for the project
+
+    Methods
+    -------
+    path_to
+
+    Examples
+    --------
+    >>> p = WnwProject()
+    >>> p.path_to(subject=1, modality="func")
+    '/data/tevesjb/wnw_project/sub-01/func'
+    """
+    def __init__(self, root: str = WNW_DEFAULT_ROOT):
+        if not isinstance(root, str):
+            raise TypeError(f"root should be str, is {type(root)}")
+        self._root = root
+    
+    def path_to(
+        self,
+        subject: int = None,
+        modality: str = 'func',
+        run: int = None,
+        acquisition: str = None,
+        echo: int = None,
+    ) -> str:
+        """Generate the path to a raw data directory or file
+
+        Parameters
+        ----------
+        subject: int
+            The subject ID
+        modality: str
+            The modality
+        run: int
+            The run number
+        acquisition: str
+            The acquisition type
+        echo: int
+            The echo number
+
+        Returns
+        -------
+        The path to the directory or file of interest
+
+        Raises
+        ------
+        TypeError, if types are not adhered to.
+        ValueError, if there are incompatible values in the call (for
+        example, requesting modality=anat and acquisition="a") OR if the
+        user requests a value which doesn't exist (for example, run=3).
+        """
+        _validate_path_to(subject, modality, run, acquisition, echo)
+        sub = f"{subject:02}"
+        path = op.join(self._root, f"sub-{sub}")
+        if modality:
+            if modality == "func":
+                path = op.join(path, "func")
+            else:
+                if modality == "anat":
+                    path = op.join(path, "anat")
+                elif modality == "T1w":
+                    path = op.join(
+                        path, "anat", f"sub-{sub}_T1w.nii"
+                    )
+        # from here, everything is functional and all parameters are used
+        if run is not None:
+            fname = (
+                f"sub-{sub}_task-wnw_acq-{acquisition}_"
+                f"run-{run}_echo-{echo}_bold.nii"
+            )
+            path = op.join(path, fname)
+
```

Comment:
Perhaps validate if the path exists before returning a value?

Reply:
The user is responsible for detecting whether the path exists; this object's purpose is to generate names only. Perhaps I could add a `validate=False` parameter to the function?
The real real reason it's like this is because the VPN was acting up and so when I wrote this the files didn't exist.

Reply:
That seems like a valid reason to omit that check.